pkg/stamper/node.go

```diff
@@ -0,0 +1,221 @@
+			continue
+		}
+
+		if !batch.Usable || batch.Utilization == 0 {
+			continue
+		}
+
+		usageFactor := batch.Depth - batch.BucketDepth              // depth - bucketDepth
+		divisor := float64(int(1) << usageFactor)                   // 2^(depth - bucketDepth)
+		stampsUsage := (float64(batch.Utilization) / divisor) * 100 // (utilization / 2^(depth - bucketDepth)) * 100
+
+		if stampsUsage >= threshold {
+			newDepth := uint16(batch.Depth) + depthIncrement
+
+			n.log.Tracef("node %s: batch %s: stamps usage %.2f%%, diluting to depth %d", n.name, batch.BatchID, stampsUsage, newDepth)
+
+			if err := n.client.Postage.DilutePostageBatch(ctx, batch.BatchID, uint64(newDepth), ""); err != nil {
+				return fmt.Errorf("node %s: dilute batch %s: %w", n.name, batch.BatchID, err)
+			}
+
+			n.log.Infof("node %s: diluted batch %s to depth %d", n.name, batch.BatchID, newDepth)
+		}
+	}
+
+	return nil
+}
+
+// Set performs both Topup and Dilute operations on postage batches in one function.
+// The order of operations is critical because Dilute increases the batch depth,
+// which directly affects the calculations for Topup by reducing the effective batch TTL.
+// Therefore, Topup is handled first, considering the original depth, followed by Dilute
+// which accounts for the new depth and utilization threshold.
+func (n *node) Set(ctx context.Context, ttlThreshold time.Duration, topupDuration time.Duration, threshold float64, depth uint16, blockTime int64, batchIds []string) error {
+	chainState, err := n.client.Postage.GetChainState(ctx)
+	if err != nil {
+		return fmt.Errorf("node %s: get chain state: %w", n.name, err)
+	}
+
+	price := chainState.CurrentPrice.Int64()
+	if price <= 0 {
+		return fmt.Errorf("node %s: invalid chain price: %d", n.name, price)
+	}
+
+	batches, err := n.client.Postage.PostageBatches(ctx)
+	if err != nil {
+		return fmt.Errorf("node %s: get postage batches: %w", n.name, err)
+	}
+
+	for _, batch := range batches {
+		if len(batchIds) > 0 && !contains(batchIds, batch.BatchID) {
+			continue
+		}
+
+		if !batch.Usable || batch.Utilization == 0 {
+			continue
+		}
+
+		// Topup
+		batchTTL := time.Unix(batch.BatchTTL, 0)
+		if time.Until(batchTTL) <= ttlThreshold {
+			originalDepth := batch.Depth - batch.BucketDepth + uint8(depth)
+			adjustedDepth := originalDepth + uint8(depth)
+			multiplier := int64(1 << adjustedDepth)
+
+			if blockTime <= 0 {
+				blockTime = 1 // avoid division by zero
+			}
+
+			remainingTTL := batchTTL.Unix() - time.Now().Unix()
+			if remainingTTL < 0 {
+				remainingTTL = 0 // handle expired batches
+			}
+
+			requiredDuration := int64(topupDuration.Seconds()) - remainingTTL
+
+			if requiredDuration > 0 {
+				amount := (requiredDuration / blockTime) * multiplier * price
+
+				n.log.Tracef("node %s: batch %s: required duration %d, amount %d", n.name, batch.BatchID, requiredDuration, amount)
+
+				if err := n.client.Postage.TopUpPostageBatch(ctx, batch.BatchID, amount, ""); err != nil {
```

Comment:
i would pass the old label to the topup request

Reply:
last argument for gas price